pkg/plugins/openai/cmd/main/main.go

```diff
@@ -0,0 +1,345 @@
+		rateLimiter: rateLimiter,
+		config:      oaiConfig,
+	}
+
+	// pluginMap is the map of plugins we can dispense.
+	var pluginMap = map[string]plugin.Plugin{
+		"CustomCostSource": &ocplugin.CustomCostPlugin{Impl: &oaiCostSrc},
+	}
+
+	plugin.Serve(&plugin.ServeConfig{
+		HandshakeConfig: handshakeConfig,
+		Plugins:         pluginMap,
+		GRPCServer:      plugin.DefaultGRPCServer,
+	})
+}
+
+func boilerplateOpenAICustomCost(win opencost.Window) pb.CustomCostResponse {
+	return pb.CustomCostResponse{
+		Metadata:   map[string]string{"api_client_version": "v1"},
+		CostSource: "AI",
+		Domain:     "openai",
+		Version:    "v1",
+		Currency:   "USD",
+		Start:      timestamppb.New(*win.Start()),
+		End:        timestamppb.New(*win.End()),
+		Errors:     []string{},
+		Costs:      []*pb.CustomCost{},
+	}
+}
+func (d *OpenAICostSource) getOpenAICostsForWindow(window opencost.Window) *pb.CustomCostResponse {
+	ccResp := boilerplateOpenAICustomCost(window)
+
+	oaiTokenUsages, err := d.getOpenAITokenUsages(*window.Start())
+	if err != nil {
+		ccResp.Errors = append(ccResp.Errors, fmt.Sprintf("error getting OpenAI token usages: %v", err))
+	}
+
+	oaiBilling, err := d.getOpenAIBilling(*window.Start(), *window.End())
+	if err != nil {
+		ccResp.Errors = append(ccResp.Errors, fmt.Sprintf("error getting OpenAI billing data: %v", err))
+	}
+
+	customCosts, err := getCustomCostsFromUsageAndBilling(oaiTokenUsages, oaiBilling)
+	if err != nil {
+		ccResp.Errors = append(ccResp.Errors, fmt.Sprintf("error converting API responses into custom costs: %v", err))
+	}
+	ccResp.Costs = customCosts
+
+	return &ccResp
+}
+
+func getCustomCostsFromUsageAndBilling(usage *openaiplugin.OpenAIUsage, billing *openaiplugin.OpenAIBilling) ([]*pb.CustomCost, error) {
+	customCosts := []*pb.CustomCost{}
+
+	tokenMap := buildTokenMap(usage)
+	for _, billingEntry := range billing.Data {
+		tokenMapKey := strings.ReplaceAll(strings.ToLower(billingEntry.Name), "-", "")
+		tokenMapKey = strings.ReplaceAll(tokenMapKey, " ", "")
+		tokenMapKey = strings.ReplaceAll(tokenMapKey, "_", "")
+
+		tokenCount, ok := tokenMap[tokenMapKey]
+		if !ok {
+			log.Debugf("no token usage found for %s", billingEntry.Name)
+			tokenCount = -1
+		}
+
+		extendedAttrs := pb.CustomCostExtendedAttributes{
+			AccountId:    &billingEntry.OrganizationID,
+			SubAccountId: &billingEntry.ProjectID,
+		}
+		customCost := pb.CustomCost{
+			BilledCost:         float32(billingEntry.CostInMajor),
+			AccountName:        billingEntry.OrganizationName,
+			ChargeCategory:     "Usage",
+			Description:        fmt.Sprintf("OpenAI usage for model %s", billingEntry.Name),
+			ResourceName:       billingEntry.Name,
+			ResourceType:       "AI Model",
+			Id:                 uuid.New().String(),
+			ProviderId:         fmt.Sprintf("%s/%s/%s", billingEntry.OrganizationID, billingEntry.ProjectID, billingEntry.Name),
+			UsageQuantity:      float32(tokenCount),
+			UsageUnit:          "tokens - All snapshots, all projects",
```

Comment:
Is there any expectation that we'll split usage by snapshot or project in the future?

Reply:
someday, when the APIs support it. As you see, there are two APIs we call: one for cost, one for usage. The only linkage we have between the two APIs is the name of the model - not the version (snapshot) of that model or the projects. 

I know that some of the other OC ecosystem maintainers are pushing OpenAI to include more fields in these APIs so that we can improve the granularity.